plugin-server/src/worker/vm/upgrades/historical-export/export-historical-events-v2.ts

```diff
@@ -0,0 +1,487 @@
+
+        let exportEventsError: Error | unknown | null = null
+
+        if (fetchEventsError) {
+            // :TODO: This doesn't stop currently running jobs from continuing
+            await meta.storage.del(EXPORT_RUNNING_KEY)
+            // :TODO: Retries logic.
+            createLog(`Failed fetching events. Stopping export - please try again later.`)
+            return
+        } else {
+            if (events.length > 0) {
+                try {
+                    await methods.exportEvents!(events)
+                } catch (error) {
+                    exportEventsError = error
+                }
+            }
+        }
+
+        if (exportEventsError instanceof RetryError) {
+            const nextRetrySeconds = 2 ** payload.retriesPerformedSoFar * 3
+
+            // "Failed processing events 0-100 from 2021-08-19T12:34:26.061Z to 2021-08-19T12:44:26.061Z. Retrying in 3s"
+            createLog(
+                `Failed processing events ${payload.offset}-${payload.offset + events.length} from ${new Date(
+                    payload.timestampCursor
+                ).toISOString()} to ${new Date(
+                    payload.timestampCursor + payload.fetchTimeInterval
+                ).toISOString()}. Retrying in ${nextRetrySeconds}s`
+            )
+
+            await meta.jobs
+                .exportHistoricalEvents({
+                    ...payload,
+                    retriesPerformedSoFar: payload.retriesPerformedSoFar + 1,
+                } as ExportHistoricalEventsJobPayload)
+                .runIn(nextRetrySeconds, 'seconds')
+        } else if (!exportEventsError) {
+            const incrementCursor = events.length < EVENTS_PER_RUN
+            const incrementedTimeCursor = Math.min(payload.endTime, payload.timestampCursor + payload.fetchTimeInterval)
+
+            await meta.jobs
+                .exportHistoricalEvents({
+                    ...payload,
+                    retriesPerformedSoFar: 0,
+                    timestampCursor: incrementCursor ? incrementedTimeCursor : payload.timestampCursor,
+                    intraIntervalOffset: incrementCursor ? 0 : payload.offset + EVENTS_PER_RUN,
+                    fetchTimeInterval: nextFetchTimeInterval(payload, events.length),
+                } as ExportHistoricalEventsJobPayload)
+                .runIn(1, 'seconds')
+        }
+        // :TODO: When unknown exportEventsError?
+
+        if (events.length > 0 && !exportEventsError) {
+            createLog(
+                `Successfully processed events ${payload.offset}-${payload.offset + events.length} from ${new Date(
+                    payload.timestampCursor
+                ).toISOString()} to ${new Date(payload.timestampCursor + payload.fetchTimeInterval).toISOString()}.`
+            )
+        }
+    }
+
+    async function getTimestampBoundaries(payload: ExportHistoricalEventsUIPayload): Promise<TimestampBoundaries> {
+        if (payload && payload.dateFrom && payload.dateTo) {
```

Comment:
I think we should actually enforce this code path here i.e. that exports are all explicitly dated. 

IIRC this is the reason the bad pattern for querying boundaries on setupPlugin arose - we were looking for a way to minimize duplication by not exporting events in the historical export that we knew the real time export will have gotten to.

Although beyond the issues we've already covered this was particularly useless on Cloud with all the deploys. 

Originally the idea was that someone would e.g. set up a plugin on self-hosted, leave it running for multiple days, then decide to do a historical export, at which point we'd grab the max timestamp determined in setupPlugin.

Either way, verbose way of saying let's simplify this and enforce boundaries? Makes code simpler and performance better. People can always set the lower bound to 1970-01-01 (at which point we should infer that that's not their lowest timestamp).

Reply:
Not sure what the suggestion here ends up being. Let's discuss the UX coming monday.

Note to frame discussion on monday that getting the boundaries is currently wrong - _timestamp != timestamp.

Reply:
TODO Make explicit.